# MiniMax
Software package MiniMax searches for groups of several enriched transcription factor binding site (TFBS) motifs in the large list of motifs obtained for the positive set of DNA sequences. The enrichment is estimated in the comparison of the test positive sequence set with the negative sequence set. Typically the positive sequences are promoters of differentially expressed genes (DEGs) derived from RNA-seq or ChIP-seq/ATAC-seq peaks. The negative sequences are taken as promoters of not-DEGs or randomly selected genomic loci, correspondingly. In the case if RNA-seq data, the DEGs/not-DEGs are defined by the criteria {adjusted p-value < 0.05 & |log2(FoldChange)| > 1} / {adjusted p-value > 0.05 & 0.8 < FoldChange) < 1.25}, respectively. For ChIP-seq/ATAC-seq data the negative set contains [randomly selected genomic loci, adopted by G/C-content selected by the AntiNoise tool](https://github.com/parthian-sterlet/antinoise/). 
