# MiniMax
Software package MiniMax searches for groups of several enriched transcription factor binding site (TFBS) motifs in the large list of motifs obtained for the positive set of DNA sequences. The enrichment is estimated in the comparison of the test positive sequence set with the negative sequence set. Typically the positive sequences are promoters of differentially expressed genes (DEGs) derived from RNA-seq or ChIP-seq peaks. 

# Description
For a motif representing TF binding sites (TFBS), the recognition accuracy is calculated as the partial area under PR curve (pAUPRC, [Levitsky et al., 2024](https://doi.org/10.18699/vjgb-24-90)); [Davis, Goadrich, 2006](https://doi.org/10.1145/1143844.1143874)). The term PR means Precision-Recall curve. Generalizing the PR curves and pAUPRC values of the several separate TFBS motifs (motif groups) to a single joint motif reveals whether the joint performance of these motifs competes with those of all participating motifs. For a single DNA sequence a hit of motif group is defined as the best hit among all motifs of this group. Testing and iptimizarion of motif content for the population of many groups by theapproach of genetic algorithm allows to detect and to rank the group of motifs most strongly reinforcing each other. The tested motifs can represent either structurally different binding sites for the same TF, or binding sites of different TFs acting as part of a single multi-protein complex (Levitsky et al., [2014](https://doi.org/10.1186/1471-2164-15-80), [2024](https://doi.org/10.18699/vjgb-24-90)). Since the default option of TFs action is mutual cooperative interactions ([Morgunova & Taipale, 2017](https://doi.org/10.1016/j.sbi.2017.03.006)), the single protein complex of multiple TFs is reflected in various portions of peaks as different enriched motifs of various TFs, including TFs participating in transcription regulation of DEGs from the RNA-seq data, and the target and partner TFs related to TFBS motif enrichment for RNA-seq or ChIP-seq data. Hence, the functional relationship of these distinct motifs is recognized through increased recognition performance in groups of enriched motifs compared to other groups of the same number of motifs. Thus, MiniMax predicts TFBS motifs respecting TFs paricipating in the same multiprotein complexes functioning in gene transcription regulation.

# Algorithm
MiniMax algorithm considers the pairs of positive/negative sequences sets derived from from RNA-seq or ChIP-seq data. peaks. The negative sequences are taken as promoters of not-DEGs or randomly selected genomic loci, correspondingly. In the case if RNA-seq data, the DEGs/not-DEGs are defined by the criteria {adjusted p-value < 0.05 & |log2(FoldChange)| > 1} / {adjusted p-value > 0.05 & 0.8 < FoldChange) < 1.25}, respectively. For ChIP-seq/ATAC-seq data the negative set contains [randomly selected genomic loci, adopted by G/C-content selected by the AntiNoise tool](https://github.com/parthian-sterlet/antinoise/). and starts from computation of two PR curves for each of two combined single motifs.
