# MiniMax
Software package MiniMax searches for groups of several enriched transcription factor binding site (TFBS) motifs in the large list of motifs obtained for the positive set of DNA sequences. The enrichment is estimated in the comparison of the test positive sequence set with the negative sequence set. Typically the positive sequences are promoters of differentially expressed genes (DEGs) derived from RNA-seq or ChIP-seq peaks. 

# Description
For a motif representing TF binding sites (TFBS), the recognition accuracy is calculated as the partial area under PR curve (pAUPRC, [Levitsky et al., 2024](https://doi.org/10.18699/vjgb-24-90); [Davis, Goadrich, 2006](https://doi.org/10.1145/1143844.1143874)). The term PR means Precision-Recall curve. Generalizing the PR curves and pAUPRC values of the several separate TFBS motifs (motif groups) to a single joint motif reveals whether the joint performance of these motifs competes with those of all participating motifs. For a single DNA sequence a hit of motif group is defined as the best hit among all motifs of this group. Testing and iptimizarion of motif content for the population of many groups by theapproach of genetic algorithm (GA) allows to detect and to rank the group of motifs most strongly reinforcing each other. The tested motifs can represent either structurally different binding sites for the same TF, or binding sites of different TFs acting as part of a single multi-protein complex (Levitsky et al., [2014](https://doi.org/10.1186/1471-2164-15-80), [2024](https://doi.org/10.18699/vjgb-24-90)). Since the default option of TFs action is mutual cooperative interactions ([Morgunova & Taipale, 2017](https://doi.org/10.1016/j.sbi.2017.03.006)), the single protein complex of multiple TFs is reflected in various portions of peaks as different enriched motifs of various TFs, including TFs participating in transcription regulation of DEGs from the RNA-seq data, and the target and partner TFs related to TFBS motif enrichment for RNA-seq or ChIP-seq data. Hence, the functional relationship of these distinct motifs is recognized through increased recognition performance in groups of enriched motifs compared to other groups of the same number of motifs. Thus, MiniMax predicts TFBS motifs respecting TFs paricipating in the same multiprotein complexes functioning in gene transcription regulation.

# Algorithm
MiniMax algorithm considers a pair of positive/negative sequence sets derived either from from RNA-seq or ChIP-seq data. peaks. For RNA-seq data the positive and negative sequences are promoters of DEGs and not-DEGs, that are defined by the criteria {adjusted p-value < 0.05 & log2(FoldChange) > 1 / log2(FoldChange) < -1  for up-/down-regulated DEGs} and {adjusted p-value > 0.05 & 0.8 < FoldChange) < 1.25}, correspondingly. For ChIP-seq data the positive/negative sequences are ChIP-seq peaksor randomly selected genomic loci, correspondingly. In the case if RNA-seq data, the , respectively. For ChIP-seq/ATAC-seq data the negative set contains [randomly selected genomic loci, adopted by G/C-content selected by the AntiNoise tool](https://github.com/parthian-sterlet/antinoise/). 

GA input data:
- a pair of positive/negative sets, containg Npos/Nneg sequences;
- a library of Mtot TFBS motifs derived from public databases like [Hocomoco](https://hocomoco14.autosome.org/) ([Vorontsov et al., 2024](https://doi.org/10.1093/nar/gkad1077)) or [Jaspar](https://jaspar.elixir.no/) ([Rauluseviciute et al., 2024](https://doi.org/10.1093/nar/gkad1059)), this library can be complemented by motifs derived by _de novo_ motifs search for an apropriate dataset, for each motifs it is required its name (e.g. TF name and motif designation) and short sescription of its TF class/family according to the hierarchical classification of mammalian ([Wingender et al., 2018](https://doi.org/10.1093/nar/gkx987)) and plant [Blanc-Mathieu et al., 2024](https://doi.org/10.1016/j.tplants.2023.06.023) TFs by DNA-binding domain (DBD) structure;
- two matrices of the size Mtot x Npos and Mtot x Nbeg. The rows of these matrices represent sequences (1 <= n <= Npos, 1 <= n <= Nneg). The m-th column (1 <= m <= Mtot) of these matrices contain -Log<sub>10</sub>(ERR) values (Expected Recognition Rate, ERR) for best predicted hits of m-th motif in the respective sequence set. 

The GA task is to obtain for these input data the list of motif groups ranked by pAUPRC accuracy, each group includes exactly M motifs, M < Mtot.

GA output data:
- a list of motif group ranked in the descending order of the pAUPRC recognition accuracy;
- a list of PR curves for these motif groups, the list is also ranked in the descending order of the pAUPRC recognition accuracy;
- a list of triangle matrices for motif groups, for positive and negative sequence sets, each matrix shows internal structure of motif groups, the matrices contains the Pearson correlation of two -Log<sub>10</sub>(ERR) values for two motifs of the same group, the list is also ranked in the descending order of the pAUPRC recognition accuracy of motif groups;
- two triangle matrices for motif groups, for positive and negative sequence sets, each matrix shows external structure of motif groups, the matrices contains the Pearson correlation of two -Log<sub>10</sub>(ERR) values for two motif groups.
